File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/interpreter/solution/Ternary.java
// Ternary.java

import java.io.PrintWriter;

public abstract class Ternary extends Operator
{
	public Ternary(Expression aOperand1, Expression aOperand2, Expression aOperand3)
	{
		setOperand1(aOperand1);
		setOperand2(aOperand2);
		setOperand3(aOperand3);
	}

	public int getNumberOfUsesOfVariable(String aName)
	{
		int result = 0;

		if (getOperand1() == null)
		{
			throw new DpException("int Ternary.getNumberOfUsesOfVariable(String aName) -- getOperand1() == null.");
		}
		result += getOperand1().getNumberOfUsesOfVariable(aName);
		if (getOperand2() == null)
		{
			throw new DpException("int Ternary.getNumberOfUsesOfVariable(String aName) -- getOperand2() == null.");
		}
		result += getOperand2().getNumberOfUsesOfVariable(aName);
		if (getOperand3() == null)
		{
			throw new DpException("int Ternary.getNumberOfUsesOfVariable(String aName) -- getOperand3() == null.");
		}
		result += getOperand3().getNumberOfUsesOfVariable(aName);

		return result;
	}

	public abstract double getValue(DpMap<String, Double> aMap);

	public abstract void output(PrintWriter aPrintWriter);

	protected void outputInfix(PrintWriter aPrintWriter, String aOperator1, String aOperator2)
	{
		aPrintWriter.print("(");
		if (getOperand1() == null)
		{
			throw new DpException("void Ternary.outputInfix(PrintWriter aPrintWriter, String aOperator1, String aOperator2) -- getOperand1() == null.");
		}
		getOperand1().output(aPrintWriter);
		aPrintWriter.print(aOperator1);
		if (getOperand2() == null)
		{
			throw new DpException("void Ternary.outputInfix(PrintWriter aPrintWriter, String aOperator1, String aOperator2) -- getOperand2() == null.");
		}
		getOperand2().output(aPrintWriter);
		aPrintWriter.print(aOperator2);
		if (getOperand3() == null)
		{
			throw new DpException("void Ternary.outputInfix(PrintWriter aPrintWriter, String aOperator1, String aOperator2) -- getOperand3() == null.");
		}
		getOperand3().output(aPrintWriter);
		aPrintWriter.print(")");
	}

	private void setOperand1(Expression aOperand1)
	{
		myOperand1 = aOperand1;
	}

	public Expression getOperand1()
	{
		return myOperand1;
	}

	private void setOperand2(Expression aOperand2)
	{
		myOperand2 = aOperand2;
	}

	public Expression getOperand2()
	{
		return myOperand2;
	}

	private void setOperand3(Expression aOperand3)
	{
		myOperand3 = aOperand3;
	}

	public Expression getOperand3()
	{
		return myOperand3;
	}

	private Expression myOperand1;
	private Expression myOperand2;
	private Expression myOperand3;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/interpreter/solution/ternary.cpp
// ternary.cpp

#include <iostream>
#include <string>
#include "dp_exception.h"
class expression;
#include "ternary.h"

ternary::ternary(expression * a_operand1, expression * a_operand2, expression * a_operand3)
{
	set_operand1(a_operand1);
	set_operand2(a_operand2);
	set_operand3(a_operand3);
}

int ternary::get_number_of_uses_of_variable(const std::string & a_name) const
{
	int result(0);

	if (get_operand1() == nullptr)
	{
		throw dp_exception("int ternary::get_number_of_uses_of_variable(const std::string & a_name) const -- get_operand1() == nullptr.");
	}
	result += get_operand1()->get_number_of_uses_of_variable(a_name);
	if (get_operand2() == nullptr)
	{
		throw dp_exception("int ternary::get_number_of_uses_of_variable(const std::string & a_name) const -- get_operand2() == nullptr.");
	}
	result += get_operand2()->get_number_of_uses_of_variable(a_name);
	if (get_operand3() == nullptr)
	{
		throw dp_exception("int ternary::get_number_of_uses_of_variable(const std::string & a_name) const -- get_operand3() == nullptr.");
	}
	result += get_operand3()->get_number_of_uses_of_variable(a_name);

	return result;
}

void ternary::output_infix(std::ostream & a_ostream, const std::string & a_operator1, const std::string & a_operator2) const
{
	a_ostream << "(";
	if (get_operand1() == nullptr)
	{
		throw dp_exception("void ternary::output_infix(std::ostream & a_ostream, const std::string & a_operator1, const std::string & a_operator2) const -- get_operand1() == nullptr.");
	}
	get_operand1()->output(a_ostream);
	a_ostream << a_operator1;
	if (get_operand2() == nullptr)
	{
		throw dp_exception("void ternary::output_infix(std::ostream & a_ostream, const std::string & a_operator1, const std::string & a_operator2) const -- get_operand2() == nullptr.");
	}
	get_operand2()->output(a_ostream);
	a_ostream << a_operator2;
	if (get_operand3() == nullptr)
	{
		throw dp_exception("void ternary::output_infix(std::ostream & a_ostream, const std::string & a_operator1, const std::string & a_operator2) const -- get_operand3() == nullptr.");
	}
	get_operand3()->output(a_ostream);
	a_ostream << ")";
}

void ternary::set_operand1(expression * a_operand1)
{
	my_operand1 = a_operand1;
}

expression * ternary::get_operand1() const
{
	return my_operand1;
}

void ternary::set_operand2(expression * a_operand2)
{
	my_operand2 = a_operand2;
}

expression * ternary::get_operand2() const
{
	return my_operand2;
}

void ternary::set_operand3(expression * a_operand3)
{
	my_operand3 = a_operand3;
}

expression * ternary::get_operand3() const
{
	return my_operand3;
}
